src/hotspot/share/classfile/fieldLayoutBuilder.cpp

```diff
@@ -758,7 +758,7 @@ void FieldLayoutBuilder::inline_class_field_sorting(TRAPS) {
             //too_volatile_to_flatten = false; //FIXME
             // volatile fields are currently never flatten, this could change in the future
           }
-          if (!(too_big_to_flatten | too_atomic_to_flatten | too_volatile_to_flatten) || fieldinfo.access_flags().is_final()) {
+          if (!(too_big_to_flatten | too_atomic_to_flatten | too_volatile_to_flatten)) {
```

Comment:
Please also add an IR framework test to check primitive class final/non-final field flattening, given that InlineTypeNode are removed during Optimization hence such a check has to be done on post Parse IR.

Reply:
The main problem for me is how to check the flatten status for an `InlineTypeNode`. Maybe we can check the decode operation generated by the field accessing (or loading) to an primitive class field. Any better idea for this?

Reply:
Couldn't you just check that there is no oop load for that particular field, i.e. no indirection to load from a heap buffer?

Reply:
If a primitive class field is flattened, then compiler will not be creating an InlineTypeNode for field access, sub-field will be part of container, isn't that sufficient to make the check, e.g.

```
primitive class ABC {
   int f1;
   long f2;
   ABC (int f1, long f2) {
      this.f1 = f1;
      this.f2 = f2;
   }
}


public class flatten {
   public final ABC field;

   public flatten(int val) {
     this.field = new ABC(val, (long)val);
   }

   public static long micro(int val) {
      flatten obj =  new flatten(val);
      return obj.field.f1 + obj.field.f2;
   }

   public static void main(String [] args) {
      long res = 0;
      for (int i = 0; i < 10000; i++) {
          res += micro(i);
      }
      System.out.println(res);
   }
}

```

We see no InlineTypeNode in post parse IR with following command line:-

-XX:CompileOnly=flatten::micro -XX:CompileCommand=compileonly,flatten::micro -XX:+EnablePrimitiveClasses -cp . flatten

But, field access does generate InlineTypeNode if flatten size explicitly set as 0
-XX:InlineFieldMaxFlatSize=0 -XX:+PauseAtStartup -XX:CompileOnly=flatten::micro -XX:CompileCommand=compileonly,flatten::micro -XX:+EnablePrimitiveClasses -cp . flatten

EA will anyways remove non-escapable allocation, but post parse IR verification should be good enough to test your changes ?

Reply:
Hi @jatin-bhateja @TobiHartmann , I'v added the IR tests which check the oop load number for non-flattened inline type fields. Please help to take a look whether it's fine. Thanks in advance!

Reply:
Thanks @XiaohongGong, LGTM!

Reply:
I also tried to check `InlineTypeNode`, but unfortunately `InlineTypeNode` is also optimized out after parsing due to no usage in this case, which is the same with the field flattened.